### PR TITLE
[Triplet Margin Loss] Issue 1118

### PR DESCRIPTION
**Pull request by cvnad1**
@vroulet Hi Vincent, Added code and tests for the Triplet Margin Loss Function #1118 . Kindly review the code and please do comment in case of any changes.

### CODE REVIEW

**Comment by cvnad1** [comment]
@vroulet May I know if there's anything that needs to be changed?

File: optax/losses/_self_supervised.py [hunk 1]
```diff
@@ -86,3 +86,79 @@ def ntxent(
...
+  metric specified by p-norm. The loss encourages
+  the distance between the anchor and positive points to be 
+  smaller than the distance between the anchor and negative 
+  points by at least the margin amount.
+
+  Args:
+      anchor: The anchor embeddings. Shape: [batch_size, feature_dim].
+      positive: The positive embeddings. Shape: [batch_size, feature_dim].
+      negative: The negative embeddings. Shape: [batch_size, feature_dim].
+      margin: The margin value. Default: 1.0.
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:125]: No need to put the default values since they are given in the signature.

File: optax/losses/_self_supervised.py [hunk 2]
```diff
@@ -86,3 +86,79 @@ def ntxent(
...
+      p: The norm degree for pairwise distance. Default: 2.
+      eps: Small epsilon value to avoid numerical issues. Default: 1e-6.
+      swap: Use the distance swap optimization from "Learning shallow
+      convolutional feature descriptors with triplet losses" 
+      by V. Balntas et al. Default: False.
+      reduction: Specifies the reduction to apply to the output:
+          'none' | 'mean' | 'sum'. Default: 'mean'.
+
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:133]: Add reference

File: optax/losses/_self_supervised.py [hunk 3]
```diff
@@ -86,3 +86,79 @@ def ntxent(
...
+  points by at least the margin amount.
+
+  Args:
+      anchor: The anchor embeddings. Shape: [batch_size, feature_dim].
+      positive: The positive embeddings. Shape: [batch_size, feature_dim].
+      negative: The negative embeddings. Shape: [batch_size, feature_dim].
+      margin: The margin value. Default: 1.0.
+      p: The norm degree for pairwise distance. Default: 2.
+      eps: Small epsilon value to avoid numerical issues. Default: 1e-6.
+      swap: Use the distance swap optimization from "Learning shallow
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:128]: Use rst formatting for references (see e.g. the docstring of Adam)

File: optax/losses/_self_supervised.py [hunk 4]
```diff
@@ -86,3 +86,79 @@ def ntxent(
...
+  positive: chex.Array,
+  negative: chex.Array,
+  *,
+  margin: float = 1.0,
+  p: int = 2,
+  eps: float = 1e-6,
+  swap: bool = False,
+  reduction: str = 'mean',
+) -> chex.Array:
+  """Triplet margin loss function.
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:112]: Add an example (doctest)

File: optax/losses/_self_supervised_test.py [hunk 5]
```diff
@@ -53,5 +53,41 @@ def test_batched(self):
     )
 
 
+class TripletMarginLossTest(chex.TestCase):
+
+  def setUp(self):
```
Comment (vroulet) [inline optax/losses/_self_supervised_test.py:58]: Avoid using numerical values as expected returns.
They may fail depending on the backend for example.
You may consider simple test cases with a "handmade" function (see e.g. the lbfgs tests). You can check for specific inputs (like zeros or ones).

You may also add a test for some specific behaviors (like using swap here).

Also you should test this function under jit/vmap etc... (see the chex.all_variant utility in some other tests).

**Review by vroulet — COMMENTED** [review]
Thank you @cvnad1 for doing this! Sorry for the delay. Here are some comments

**Comment by Saanidhyavats** [comment]
@vroulet we have worked on your suggestion and all the tests are passing. I think the code is ready to be merged.

File: optax/losses/_self_supervised.py [hunk 6]
```diff
@@ -86,3 +86,95 @@ def ntxent(
...
+      swap: Use the distance swap optimization
+      reduction: Specifies the reduction to apply to the output:
+          'none' | 'mean' | 'sum'
+
+  Returns:
+      The triplet margin loss value.
+      If reduction is 'none': tensor of shape [batch_size]
+      If reduction is 'mean' or 'sum': scalar tensor.
+  
+  example:
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:142]: Use `Examples` instead of example
Use 
```
optax.losses.triplet_margin_loss
```

File: optax/losses/_self_supervised.py [hunk 7]
```diff
@@ -86,3 +86,95 @@ def ntxent(
...
+  *,
+  margin: float = 1.0,
+  p: int = 2,
+  swap: bool = False,
+  eps: float = 1e-6,
+  reduction: str = 'mean',
+) -> chex.Array:
+  """Triplet margin loss function.
+
+  References:
```
Comment (vroulet) [inline optax/losses/_self_supervised.py:114]: Put references at the end of the docstring

File: optax/losses/_self_supervised_test.py [hunk 8]
```diff
@@ -53,5 +53,69 @@ def test_batched(self):
...
+    super().setUp()
+    self.a1 = jnp.ones((2, 2))
+    self.p1 = jnp.zeros((2, 2))
+    self.n1 = jnp.ones((2, 2))*2
+    self.a2 = jnp.zeros((2, 2))
+    self.p2 = jnp.ones((2, 2))
+    self.n2 = jnp.ones((2, 2))*2
+
+  @chex.all_variants
+  def test_batched(self):
```
Comment (vroulet) [inline optax/losses/_self_supervised_test.py:68]: Use parameterized tests

File: optax/losses/_self_supervised_test.py [hunk 9]
```diff
@@ -53,5 +53,69 @@ def test_batched(self):
...
+                    self.a1, self.p1, self.n1)
+
+    # JIT compiled function result
+    jit_loss = (self.variant(jax.jit(
+               _self_supervised.triplet_margin_loss))
+               (self.a1, self.p1, self.n1))
+    np.testing.assert_allclose(jit_loss, original_loss,
+                               atol=1e-4)
+
+    # VMAP applied function result
```
Comment (vroulet) [inline optax/losses/_self_supervised_test.py:109]: Make separate test for vmap

File: optax/losses/_self_supervised_test.py [hunk 10]
```diff
@@ -53,5 +53,69 @@ def test_batched(self):
...
+    handmade_result = testing_triplet_loss(
+    a=self.a2, p=self.p2, n=self.n2, margin=1.0, swap=True)
+    result = self.variant(_self_supervised.triplet_margin_loss)(
+    anchor=self.a2, positive=self.p2, negative=self.n2, swap=True)
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+    original_loss = _self_supervised.triplet_margin_loss(
+                    self.a1, self.p1, self.n1)
+
+    # JIT compiled function result
+    jit_loss = (self.variant(jax.jit(
```
Comment (vroulet) [inline optax/losses/_self_supervised_test.py:103]: No need to test jit if you already used self.variant.
self.variant is either the function as is or its jit version etc...

**Review by vroulet — COMMENTED** [review]
Thanks! There is still a bit of work. I added a few comments.
In addition:
- add the loss to the __init__ file of the losses folder
- add the function to the api docs

**Comment by cvnad1** [comment]
@vroulet 
Apologies for the late response but for the past couple of days we have been trying to implement your suggestions and were facing some errors which we rectified for most things. In the latest commit, we are facing some new errors which we are unable to comprehend. It would be of great help if you could review our latest commit and give your valuable feedback to solve this. Again, Sorry for all the commits we made with errors.

**Comment by Saanidhyavats** [comment]
@vroulet we tried multiple things to solve the error in pipeline, the tests for triplet_loss are passing locally. The errors we are getting here seems to be not from the function we implemented. Can you guide us on this?

**Comment by vroulet** [comment]
Hello @cvnad1 , @Saanidhyavats ,
Yes, a recent submit had broken the tests at head, it's been fixed. You can sync with head and rerun.

**Review by vroulet — COMMENTED** [review]
Last round of comments, I'll take care of further formattings on my end.
Could you also squish your commits if possible? Once the PR will be merged this will facilitate pinpointing the changes rather than having 57 commits for one PR.

**Comment by Saanidhyavats** [comment]
@vroulet we have worked on the suggested changes. It will be really helpful if you can squash commits for us. Thank you!

**Comment by rdyro** [comment]
If you can introduce the remaining changes @vroulet asked for, I can merge this PR for you. Thanks for this great contribution!

**Comment by cvnad1** [comment]
@rdyro I believe we accounted for everything. Could you kindly tell if we are missing on something?

**Comment by rdyro** [comment]
> @rdyro I believe we accounted for everything. Could you kindly tell if we are missing on something?

Oops, I was looking at an old snapshot, thanks, looks good!

**Review by rdyro — APPROVED** [review]
